cmd/kubeadm/app/master/manifests.go

```diff
@@ -367,11 +393,12 @@ func getAPIServerCommand(cfg *kubeadmapi.MasterConfiguration, selfHosted bool, k
 	}
 
 	// Check if the user decided to use an external etcd cluster
-	if len(cfg.Etcd.Endpoints) > 0 {
-		command = append(command, fmt.Sprintf("--etcd-servers=%s", strings.Join(cfg.Etcd.Endpoints, ",")))
-	} else {
-		command = append(command, "--etcd-servers=http://127.0.0.1:2379")
+	var etcdIPs string
+	if selfHosted {
+		etcdIPs += fmt.Sprintf("http://%s:2379,", cfg.Etcd.Cluster.ServiceIP)
```

Comment:
I was also going to implement TLS in a follow-up PR because this one's already quite large. If it's considered enough of a priority, I can add to this one, however. What do you think?

Basically if a user provided TLS certs for etcd, use them, otherwise generate them.